### ISSUE TITLE
Module version mismatch with node-canvas on AWS Lambda 

### ISSUE THREAD

**Issue by LucianoVandi** [serverless/serverless#4516]
# This is a Bug Report

## Description
I'm trying to test a node express app with the following dependencies:

```
"dependencies": {
    "aws-serverless-express": "^3.0.2",
    "canvas": "^1.6.7",
    "express": "^4.16.2",
    "gifencoder": "^1.0.6",
    "moment": "^2.13.0",
    "serverless-apigw-binary": "^0.4.4"
  }
```

Since i'm working on OSX, I run `npm install --production` on a docker container with the official aws image (**amazonlinux:2017.03.1.20170812**) and **node v6.10.3** as suggested on AWS Lambda docs. The build process completes without errors.

After the image is build i copy node_modules content to the host directory with `docker cp {id}:/app/node_modules .`

Then from the osx ship the app with `serverless deploy`. 

After the process completes, `serverless invoke -f myfunction` returns:

```
{
    "errorMessage": "Module version mismatch. Expected 48, got 57.",
    "errorType": "Error",
    "stackTrace": [
        "Object.Module._extensions..node (module.js:597:18)",
        "Module.load (module.js:487:32)",
        "tryModuleLoad (module.js:446:12)",
        "Function.Module._load (module.js:438:3)",
        "Module.require (module.js:497:17)",
        "require (internal/module.js:20:19)",
        "Object.<anonymous> (/var/task/node_modules/canvas/lib/bindings.js:3:18)",
        "Module._compile (module.js:570:32)",
        "Object.Module._extensions..js (module.js:579:10)"
    ]
}
```

From this error it seems that canvas was compiled with node_module_version=57, that correspond to Node >= 8.0.0... but I'm sure it get compiled with v6.10.3. What am i doing wrong?

## Additional Data

* ***Serverless Framework Version you're using***: 1.17.0
* ***Operating System***: OS X 10.11.6

**Comment by LucianoVandi** [comment]
I've tried to dig deeper and found that in `node_modules/canvas/build/config.gypi` there are the following config values:

```
"node_module_version": 48,
"node_version": "6.10.3",
```
So where does it get module version 57? For what I know serverless does not run `npm install`, but just creates a zip and upload to S3. Am I wrong? Should I need to install the same version of node on my host machine?

**Comment by HyperBrain** [comment]
@LucianoVandi You might be right. If any module requires local compilation if it is installed (like the gyp file shows), it will of course be compiled with and for the current system and node version you are using. In your case, if there are any restrictions for the compiled code, they will target the Node version used.

To prove if my statement is right, you'd just do the same thing with Node 6 and see if the problem is gone.

Even more critical would be, if you have C/C++ (native) modules compiled as node modules - then even the OS would matter and you'd need that to do on an AWS Amazon Linux.

**Comment by LucianoVandi** [comment]
Thanks @HyperBrain I've installed Node v6.10.3 locally and now the error "module mismatch" disappeared. So I'm assuming that `serverless deploy` is compiling locally (on darwin) before uploading the zip file.

I really miss something... I'm doing `npm install` on the docker container with the same architecture declared in the [Lambda docs](http://docs.aws.amazon.com/lambda/latest/dg/current-supported-versions.html) to avoid conflicts.

**Comment by HyperBrain** [comment]
@LucianoVandi Found an interesting thread about canvas and AWS Lambda. Seems it is a bit tricky 😄 
https://github.com/Automattic/node-canvas/issues/680

**Comment by LucianoVandi** [comment]
Hi @HyperBrain I've finally found a solution by following this [guide](https://github.com/Automattic/node-canvas/wiki/Installation---Amazon-Linux-AMI-(EC2)) with some small modification.

The build process is run on a docker container, this should ensure the environment is really similar to AWS Lambda. I'm still investigating on a issue with pango... Running the function throws "libpangocairo cannot open shared object file".

Looking in `canvas/binding.gyp` there is a configuration value with this content 
```
...
'with_pango%': '<!(./util/has_lib.sh pango)',
...
```
Setting this value to false make everything work fine. As soon as I have solved it, I will post a link to the Dockerfile and a shell script I'm using to build, copy node_modules from docker and deploy with serverless. Thanks for pointing me the right way! ;)
